### PR TITLE
Ketcher2 Common Templates, User Templates & User settings

### PR DESCRIPTION
**Pull request by haditariq**
Purpose of this PR is to resolve [Issues # 191](https://github.com/ComPlat/chemotion/issues/191) , [Issues # 199](https://github.com/ComPlat/chemotion/issues/199) & [Issue # 192](https://github.com/ComPlat/chemotion/issues/192)

1 - (Issue 191) User is provided with a list of global/common templates to use in the ketcher2.
 ***Note: common templates is not complete as you can see when a list of templates is loaded it has a dummy display. That can be fixed using indigo service which is implemented is the other PR so on merge it can be resolved.
 
https://github.com/user-attachments/assets/91a656f6-04e3-438f-aca6-ddfe6944b407

2 - (Issue 199) Ketcher2 setting are stored for each user as preferences for later effects.

https://github.com/user-attachments/assets/2af9ff08-c2c8-48b9-8574-0b50e76d81f6

3 - (Issue 192) User templates available to be used within Ketcher2.

https://github.com/ComPlat/chemotion_ELN/assets/25753582/612006a5-0886-476c-8357-9c28e8046169

---------
- [ ] rather 1-story 1-commit than sub-atomic commits

- [ ] commit title is meaningful =>  git history search

- [ ] commit description is helpful => helps the reviewer to understand the changes

- [ ] code is up-to-date with the latest developments of the target branch (rebased to it or whatever) => :fast_forward:-merge for linear history is favoured

- [ ] added code is linted

- [ ] tests are passing (at least locally): we still have some random test failure on CI. thinking of asking spec/examples.txt to be commited

- [ ] in case the changes are visible to the end-user,  video or screenshots should be added to the PR => helps  with user testing

- [ ] testing coverage improvement is improved.

- [ ] CHANGELOG :  add a bullet point on top (optional: reference to github issue/PR )

- [ ] parallele PR for documentation  on docusaurus  if the feature/fix is tagged for a release

### CODE REVIEW

File: app/api/chemotion/profile_api.rb [hunk 1]
```diff
@@ -125,17 +141,85 @@ class ProfileAPI < Grape::API
...
+      params do
+        requires :path, type: String, desc: 'file path of user template'
+      end
+      delete do
+        user_templates = current_user.profile.user_templates
+        user_templates.delete(params[:path])
+
+        # remove file from store
+        file_path = Rails.root.join('uploads', Rails.env, params[:path])
+        File.delete(file_path) if File.exist?(file_path)
```
[the remark this answers is not in the document]
Comment (adambasha0) [reply app/api/chemotion/profile_api.rb:197]: use FileUtils.rm_f(file_path) instead

**Review by adambasha0 — CHANGES_REQUESTED** [review]
- Eslint warnings
- Spec tests for new APIs
- SurfaceChemistry and unused func. into a separate PR
- Documentation for the feature on [Docusaurs](https://chemotion.net/docs/documentation)

File: app/packs/src/fetchers/CommonTemplateFetcher.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import 'whatwg-fetch';
+
+export default class CommonTemplatesFetcher {
```
[the remark this answers is not in the document]
Comment (haditariq) [reply app/packs/src/fetchers/CommonTemplateFetcher.js:3]: Used at app/packs/src/components/structureEditor/StructureEditorModal.js#L415

File: db/seeds/shared/matrice.seed.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+attributes = {
```
[the remark this answers is not in the document]
Comment (haditariq) [reply db/seeds/shared/matrice.seed.rb:1]: Changed to: ketcher2_matrice.seed.rb

File: app/packs/src/components/ketcher-templates/CommonTemplateItem.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+
+function CommonTemplateItem(props) {
+  const { item, onClickItem } = props;
+  let iconPath = '/images/ketcherails/icons/small/';
```
[the remark this answers is not in the document]
Comment (haditariq) [reply app/packs/src/components/ketcher-templates/CommonTemplateItem.js:6]: It is required for later use with indigo service.

File: app/api/chemotion/profile_api.rb [hunk 5]
```diff
@@ -64,12 +65,27 @@ class ProfileAPI < Grape::API
           data[dt] = sorted_layout
         end
 
+        if profile.user_templates?
+          profile.user_templates.each do |template|
```
[the remark this answers is not in the document]
Comment (haditariq) [reply app/api/chemotion/profile_api.rb]: This part of code has changed with the implementation approach.

File: app/api/chemotion/profile_api.rb [hunk 6]
```diff
@@ -125,17 +141,85 @@ class ProfileAPI < Grape::API
...
+
+      desc 'post user template'
+      params do
+        requires :content, type: String, desc: 'ketcher file content'
+      end
+      post do
+        file_name = 'template.txt'
+        file_path = Rails.root.join('uploads', Rails.env, file_name)
+        begin
+          File.new(file_path, 'w') unless File.exist?(file_path)
```
[the remark this answers is not in the document]
Comment (haditariq) [reply app/api/chemotion/profile_api.rb:163]: This part of code has changed with the implementation approach.

File: app/api/chemotion/profile_api.rb [hunk 7]
```diff
@@ -125,17 +141,85 @@ class ProfileAPI < Grape::API
...
+        Ketcher2Setting.find_by(user_id: current_user.id)
+      end
+
+      desc 'update user profile editor ketcher 2 setting options'
+      params do
+        requires :data, type: String, desc: 'data structure for ketcher options'
+      end
+      put 'editors/ketcher2-options' do
+        Ketcher2Setting.where(user_id: current_user.id).delete_all
+        new_settings = Ketcher2Setting.create({ user_id: current_user.id }.merge(JSON.parse(params[:data])))
```
[the remark this answers is not in the document]
Comment (haditariq) [reply app/api/chemotion/profile_api.rb]: This part of code has changed with the implementation approach.

**Comment by haditariq** [comment]
- Eslint warnings: Done
- SurfaceChemistry and unused func. into a separate PR: Done

- Documentation for the feature on [Docusaurs](https://chemotion.net/docs/documentation):  Please share details on how it can be done.

User templates and ketcher user settings are re-implemented (usage still same). I'm looking forward to your next review.


@adambasha0

**Review by adambasha0 — APPROVED** [review]

File: 4.json [hunk 8]
```diff
@@ -0,0 +1 @@
+{"resetToSelect":"paste","rotationStep":15,"showValenceWarnings":true,"atomColoring":false,"showStereoFlags":true,"stereoLabelStyle":"Iupac","colorOfAbsoluteCenters":"#ff0000","colorOfAndCenters":"#0000cd","colorOfOrCenters":"#228b22","colorStereogenicCenters":"LabelsOnly","autoFadeOfStereoLabels":true,"absFlagLabel":"ABS","andFlagLabel":"AND Enantiomer","mixedFlagLabel":"Mixed","ignoreChiralFlag":false,"orFlagLabel":"OR Enantiomer","font":"30px Arial","fontsz":13,"fontszsub":13,"carbonExplicitly":false,"showCharge":true,"showValence":true,"showHydrogenLabels":"Terminal and Hetero","aromaticCircle":true,"doubleBondWidth":6,"bondThickness":2,"stereoBondWidth":6,"dearomatize_on_load":false,"smart_layout":true,"ignore_stereochemistry_errors":true,"mass_skip_error_on_pseudoatoms":false,"gross_formula_add_rsites":true,"gross_formula_add_isotopes":true,"showAtomIds":false,"showBondIds":false,"showHalfBondIds":false,"showLoopIds":false,"miewMode":"LN","miewTheme":"light","miewAtomLabel":"bright"}
```
Comment (phuang26) [inline 4.json:1]: file to be deleted

File: READ_ME_Ketcher2.md [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    sudo apt install jq
+    ```
+
+## Installation Steps
+
+### 1. Add Ketcher Files to `public/`
+
+- Run the installation script:
+    ```bash
+    bin/ket2-install.sh
```
Comment (phuang26) [inline READ_ME_Ketcher2.md:18]: change to chem-ket2-install.sh

File: app/assets/images/svg_icons/shapes.svg
Comment (phuang26) [inline app/assets/images/svg_icons/shapes.svg:1]: where is this svg being used?

File: app/models/ketcher2_setting.rb
Comment (phuang26) [inline app/models/ketcher2_setting.rb:1]: to be deleted

File: db/seeds/shared/ketcher2_matrice.seed.rb [hunk 10]
```diff
@@ -0,0 +1,10 @@
+attributes = {
```
Comment (phuang26) [inline db/seeds/shared/ketcher2_matrice.seed.rb:1]: to be removed

File: db:seed [hunk 11]
```diff
@@ -0,0 +1,11 @@
+=> [[32m#<Matrice:0x0000555663b89db8[0m
```
Comment (phuang26) [inline db:seed:1]: to be removed

File: public/safety_sheets/131377_Merck.pdf
Comment (phuang26) [inline public/safety_sheets/131377_Merck.pdf:1]: to be removed

File: public/safety_sheets/252549_Merck.pdf
Comment (phuang26) [inline public/safety_sheets/252549_Merck.pdf:1]: to be removed

File: READ_ME_Ketcher2.md (hunk 9 above)
Comment (haditariq) [reply READ_ME_Ketcher2.md:18]: done

File: app/assets/images/svg_icons/shapes.svg
Comment (haditariq) [reply app/assets/images/svg_icons/shapes.svg:1]: File removed, not used.

File: app/models/ketcher2_setting.rb
Comment (haditariq) [reply app/models/ketcher2_setting.rb:1]: done

File: db/seeds/shared/ketcher2_matrice.seed.rb (hunk 10 above)
Comment (haditariq) [reply db/seeds/shared/ketcher2_matrice.seed.rb:1]: done

File: public/safety_sheets/131377_Merck.pdf
Comment (haditariq) [reply public/safety_sheets/131377_Merck.pdf:1]: done

File: db:seed (hunk 11 above)
Comment (haditariq) [reply db:seed:1]: done

File: public/safety_sheets/252549_Merck.pdf
Comment (haditariq) [reply public/safety_sheets/252549_Merck.pdf:1]: done

**Pull request merged**